Block Store Addition

Olshansk opened the pull request:
## Description
Addition of block store to V1.

Fixes https://github.com/pokt-network/pocket/issues/120

## Type of change
Please mark the options that are relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## List of Changes

**Tl;dr summary of changes**
- Add a key-value block store
- Hydrate the database with the genesis state
- Load state from a previous height when the nodes startup

**Persistence - blockStore specific changes**
* Block store path: Add `block_store_path` to the persistence configs
* Genesis DB: Properly hydrate the Postgres DB based on the genesis state
* KVStore: Use `badgerDB` for the KVStore

**Persistence - block store supporting changes**
* Add `PromptShowLatestBlockInStore` to the debug client
* Add the `node config` to the `bus` so it is accessible throughout the entire node
* Use the real persistence module in `LocalNet` even though PrePersistence is not deprecated yet
* Removed `PostgresDB` struct to simplify logic
* Storing the transaction in the persistence context
* Allow the persistence module to handle “debug” messages

**Flyby changes - outside of persistence**
* Extract the utility context specific functionality into a separate file
* Rename `end_height` to `height` in the gov params schema
* Rename some functions in the persistence / consensus / utility modules to make the code easier to read
    * E.g. `GetTransactionsForProposal` to `GetProposalTransactions`
    * E.g. `ApplyBlock` to `ApplyProposalTransactions`

**Plans for next commit(s)**
- [ ] Schema validation
- [ ] Computation of state hash via Merkle tree 
- [ ] Add sequence diagrams for end-to-end flows in the READMEs

**Followup TODOs - Simple**
- [ ] Since the `node config` is available via the `bus`, remove local storage/copies of it in other modules
- [ ] Discuss if `ApplyProposalTransactions` going to return `blockHash` or `appHash`?
- [ ] Properly access and cleanup the mempool

**Followup TODOs - Complex**
- [ ] End-to-end unit tests for loading / checking preloaded state
- [ ] Discuss how we can have multiple parallel isolated context if everyone is writing to the same DB
- [ ] Discuss whether we should be using a KV store at all (to be discussed in protocol hour)
- [ ] Discuss if BadgerDB is okay or not?
    - Fully implemented in Go
    - Embedded lightweight KV Store
    - More performant than RocksDB (i.e. an improved fork of LevelDB)
    - Touted by various online threads
    - LSM tree (good for writes) based KV Store rather than B+ tree (good for reads, such as hstore)
    - Decision not final, just quick and easy

## How Has This Been Tested?
- [X] `make test_all`
- [X] LocalNet

Olshansk committed:
Things kind of working again but had to change a lot of things

Olshansk committed:
Minor cleanup

Olshansk committed:
Self review with DISCUSS_IN_THIS_COMMIT additions

Olshansk commented:
I took a first stab at hacking on this (still only like 10%-20%) of the way there, but I wanted to jump on a team call next Monday and go over the `DISCUSS_IN_THIS_COMMIT` comments I left to get your thoughts & ideas (@andrewnguyen22) before I spend more time on this PR.

cc @phthan0 for context

Olshansk committed:
Added CommitPersistenceContext

Olshansk committed:
Removed PostgresDB

Olshansk committed:
Separated context and block operations in persistence

Olshansk committed:
Added TransactionExists

Olshansk committed:
Rough work to commit nlocks

Olshansk committed:
Updated Dockerfile with badger CLI

Olshansk committed:
Added ShowLatestBlock from BlockStore

phthan0 commented on persistence/context.go:
```diff
@@ -0,0 +1,36 @@
+
+func (p PostgresContext) AppHash() ([]byte, error) {
+	log.Println("TODO: AppHash not implemented")
+	return []byte("A real app hash, I am not"), nil
```
lol

phthan0 commented on utility/transaction.go:
```diff
@@ -79,6 +84,7 @@ func (u *UtilityContext) GetProposalTransactions(proposer []byte, maxTransaction
 	return transactions, nil
 }
 
+// Exposed for testing purposes only
```
famous last words

Olshansk replied:
Adding the `CLEANUP` tag :)

phthan0 reviewed (approved):
Not a lot of controversial stuff in this to my eye. LGTM

Olshansk commented:
Sounds good.

It's still a WIP. I just hacked together all the connections and haven't added tests, documentation or cleaned up the code yet.

Will request a review once those are done.

Olshansk committed:
Minor cleanup after self review

Olshansk commented:
Didn't have time to work on this today at all, but figured I'd share some of the bullet points I made in terms of my thinking, summarizing the changes, etc...

This will be reflected in the issue description upon completion.

```
In next commit
- Merkle tree
- Hash of validator
- Schema guarantee
- Review the end-to-end clarity of the algorithm

In this commit
- Load from previous state (KVStore & DB) if possible
- Panic if more than one context
- Move from temp KV Store to permanent KV Store
- Write to the DB
- Include this: https://mermaid.live/edit#pako:eNp1Vdtu2zAM_RXCQIEWa1Bgj8FQoFs7oMgF3pJlL3lhLCYRKkueJKfNiv77SDmJncv8YoM6OjzkoeT3rHCKsn4W6E9NtqBHjSuP5dwCPxX6qAtdoY3wUFWAQV5GFxi1s9fh5hyWzwSVu-KFIszQaIXR-YvQUS7Q6RuMqMydM-eIyY_h41cB8Qc8YsQFBjqHDUeCGboCjXA5v72QbCqYEfkXQzD1dIFmkKQPaNtj3TXBhIVfgOXjToVjiq_Ov8xtg7u6gu_aJBXVoSLjXAW5d5VbYeTkbwHi2rt6teY3Qf45b2lg93Cbe737-3zWZ3wbzmccHOVtkKzq5B4SKvJAhgoxqImPHSd1G44L2yQtkpIaWINVYNKmPQvyzo3IzGcd4p-E1XFRScknkXK-Nsp5rSfZhjpEcEupec-WurEbDIInLNZsB9qAHcltgjQBffiy8PftUgruU3B6BZYKCgH9FpgVYek8xNO2CfiQN_43pcCebWFqJVaBtrAw7DVUYmBAcwLfCfztNbOq3YiCXsJGUnVMOrdiidqwER0pAdAT22IkvsCiO1cT8pop_9K-_yD9GY66zeGD0LYlek0bgkLYEjuEYk0lhi5BMxI7ZqgraY66Q6VYQIi-LmLt6di7Xwl092wp-AgkBh6Q3eYkfaOpdDMQI9mThl8mbtki-WQKcjA7tXmQKHpCkY4iXLMufwthr1eJKdHdnHKlBrRcnTOSCpDpSLI7rWGbU0gcDBQv1nGsrpPq0OuyuV68c_Fw43Wyty3_5sqq5k6EKP1YY1gLOVqpiCoZgjR07cZ0GYz7zTUSdssQHU9-5-aw7XyNU5JSx3THWHo9ZlTUOejZbcbKS9SKfwTvAphnvKukedbnT4XMn902cUnxYPTKNmuGlnGeze0HczTuPinNZmX9JZpAtxnW0U22tsj6PCO0B-3-MjvUxz-4MBGO

Persistence - BlockStore specific changes
* Block store path: Add `block_store_path` to the persistence configs
* Genesis DB: Properly hydrate the Postgres DB based on the genesis state
* KVStore: Use `badgerDB` for the KVStore
* StoreBlock
    * Added `StoreBlock` to the utility module
    * 
    * Storing the block 

Persistence - supporting changes
* Using the real persistence module in `LocalNet` but not deprecated yet
* Add `PromptShowLatestBlockInStore` to the debug client
* Add the `node config` to the `bus` so it is accessible throughout
* Added badger CLI to the Dockerfile
* Removed `PostgresDB` struct to simplify logic
* Storing the transaction in the persistence context

Flyby changes
* Rename some utility specific methods so code is easier to read
    * E.g. `GetTransactionsForProposal` to `GetProposalTransactions`
    * E.g. `ApplyBlock` to `ApplyProposalTransactions`
* Rename `end_height` to `height` in the gov params schema

Followup TODOs - Simple
- [ ] Since the `node config` is available via the `bus`, remove local storage/copies of it in other modules
- [ ] Discuss if `ApplyProposalTransactions` going to return `blockHash` or `appHash`?
- [ ] Properly access and cleanup the mempool

Followup TODOs - Complex
- [ ] Discuss how we can have multiple parallel isolated context if everyone is writing to the same DB
- [ ] Discuss whether we should be using a KV store at all (see supporting sequence diagram)
- [ ] Discuss if BadgerDB is okay or not?
    - Fully implemented in Go
    - Embedded lightweight KV Store
    - More performant than RocksDB (i.e. an improved fork of LevelDB)
    - Touted by various online threads
    - LSM tree (good for writes) based KV Store rather than B+ tree (good for reads, such as hstore)
    - Decision not final, just quick and easy

Test passing
* `make test_all`
* `LocalNet`
```

Olshansk committed:
Tests passing again

Olshansk committed:
Implemented writing blocks to DB - no new tests yet

Olshansk committed:
Write to SQL DB successful

Olshansk committed:
Temp commit - things not really functional but getting there

Olshansk committed:
Merge with main

Olshansk committed:
Fixed issues after merging with main

Olshansk committed:
Temporarily unblocked persistence and consensus modules from starting so all tests pass

Olshansk committed:
Consensus tests passing again

Olshansk committed:
All unit tests passing again

Olshansk committed:
Completed a self review of the code

Olshansk committed:
Loose ends needed for LocalNet to work and create a demo video

Olshansk commented:
@andrewnguyen22 PTAL when you have a chance.

Consider it a first iteration since I already made a significant amount of changes and want to reduce the size/scope/impact of this PR while adding new baseline functionality. 
* I created a short demo video [here](https://drive.google.com/file/d/1F0zyp3nlHruOHloJLymmutzHpSduz1JZ/view?usp=sharing), but did not put time into editing (speed modifications) or turning it into a gif, so it's a bit long.
* Please also read the issue description where I detailed out the changes in this PR, as well as next steps.

andrewnguyen22 commented on persistence/kvstore/kvstore.go:
```diff
@@ -0,0 +1,93 @@
+import (
+	"log"
+
+	badger "github.com/dgraph-io/badger/v3"
```
Why badger-db?

Leveldb would've been my first choice as far a placeholder considering it's already used in pre-persistence.

Olshansk replied:
Discussed offline. Leaving as is for now and we'll figure it out.

andrewnguyen22 commented on persistence/pre_persistence/persistence.go:
```diff
@@ -266,9 +279,21 @@ func (m *PrePersistenceContext) GetBlockHash(height int64) ([]byte, error) {
+	return db.Contains(append(TransactionKeyPrefix, []byte(transactionHash)...)), nil
+}
+
+func (m *PrePersistenceContext) StoreTransaction(transactionProtoBytes []byte) error {
```
So you combined the blockstore with the transaction indexer.

I haven't seen this done in this way before.

I think you need to be explicit about this in the issue

andrewnguyen22 replied:
Aka expand the scope or we're gonna see double work

Olshansk replied:
`Aka expand the scope or we're gonna see double work`

My personal opinion is that it's always better to reduce the scope and simply do the work in multiple PRs if possible.

I removed it from this commit and moved those changes over to `issues/120/blockstore/txindexer`.

andrewnguyen22 replied:
Okay sounds good

andrewnguyen22 commented on utility/block.go:
```diff
@@ -17,26 +17,32 @@ DeliverTx individually applys each transaction against the state and rolls it ba
 */
 
-func (u *UtilityContext) ApplyBlock(latestHeight int64, proposerAddress []byte, transactions [][]byte, lastBlockByzantineValidators [][]byte) ([]byte, error) {
+func (u *UtilityContext) ApplyProposalTransactions(latestHeight int64, proposerAddress []byte, transactions [][]byte, lastBlockByzantineValidators [][]byte) ([]byte, error) {
```
I'm not sold on this name change. The lifecycle includes 'begin_block' and 'end_block' which is completely separate from 'transactions'

Olshansk replied:
Reverting back to ApplyBlock.

Olshansk committed:
Rename maybeLoadState to loadPersistedState

Olshansk commented on consensus/module.go:
```diff
@@ -138,6 +143,32 @@ func (m *consensusModule) SetBus(pocketBus modules.Bus) {
 	m.leaderElectionMod.SetBus(pocketBus)
 }
 
+func (m *consensusModule) maybeLoadState() error {
```
Renamed it from `maybeLoadState` to `loadPersistedState`.

Please resolve this if that's better or if you think it should just be `loadState`

Olshansk commented on consensus/module.go:
```diff
@@ -153,7 +184,7 @@ handler which has both pros and cons:
 	Cons:
 		* The leader's "replica related logic" requires an additional code path
-		* Code is less "generalizable" and therefore potentially more error prone
+		* Code is less "[gene]ralizable" and therefore potentially more error prone
```
Don't know how this happened. Reverting.

Olshansk commented on consensus/hotstuff_leader.go:
```diff
@@ -41,7 +41,7 @@ func (handler *HotstuffLeaderMessageHandler) HandleNewRoundMessage(m *consensusM
 	// TODO(olshansky): Add more unit tests for these checks...
 	if highPrepareQC == nil || highPrepareQC.Height < m.Height || highPrepareQC.Round < m.Round {
-		block, err := m.prepareBlock()
+		block, err := m.prepareBlockAsLeader()
```
Thank you

Olshansk commented on persistence/db.go:
```diff
@@ -129,17 +140,126 @@ func InitializeAccountTables(ctx context.Context, db *pgx.Conn) error {
+	if err := ctx.InitParams(); err != nil {
+		return err
+	}
+	// TECHDEBT: Calling this for each param is not practical. Need someone to refactor the gov params table.
```
Updated the comment to this:
```
	// TECHDEBT: Calling the setter (e.g. SetValidatorMaximumMissedBlocks, Set_) for this for each param results in a lot
	// of code bloat and is not practical. This will need to be fixed as part of the gov param refactor task.
```

Lmk if that's clearer.

Olshansk committed:
Reverted ApplyProposalTransactions to ApplyBlock

Olshansk commented on consensus/types/errors.go:
```diff
@@ -142,7 +142,7 @@ const (
 	commitBlockError                            = "could not commit block"
 	replicaPrepareBlockError                    = "node should not call `prepareBlock` if it is not a leader"
-	leaderErrApplyBlock                         = "node should not call `applyBlock` if it is leader"
+	leaderErrApplyProposalTransactions          = "node should not call `applyBlock` if it is leader"
```
Reverting to what we had. Not a blocker and not a strongly held opinion.

Olshansk committed:
Remove tx indexer logic

Olshansk commented:
@andrewnguyen22 PTAL when you have a chance. I left the comments that may need a second look/confirmation from you unresolved.
